TTS/enhancer/models/bwe.py

```diff
@@ -0,0 +1,308 @@
+                i == 0
+            ) for i in range(len(self.kernel_sizes))
+        ])
+        self.last_conv = nn.Conv2d(32, 1, (15,5), (15,1))
+
+    def forward(self, x):
+        feats = []
+        x = self.mel_spec(x).transpose(1,2).unsqueeze(1)
+        for disc in self.layers:
+            x = disc(x)
+            feats.append(x)
+        x = self.last_conv(x)
+        feats.append(x)
+        x = torch.flatten(x, 1, -1)
+        return x, feats
+
+class BWE(BaseTrainerModel):
+    def __init__(
+        self,
+        config: Coqpit,
+        ap: "AudioProcessor",
+    ):
+        super().__init__()
+        self.config = config
+        self.ap = ap
+        self.args = config.model_args
+        self.input_sr = config.input_sr
+        self.target_sr = config.target_sr
+        self.scale_factor = (self.target_sr / self.input_sr, )
+        self.train_disc = False
+
+        self.upsample = Upsample(scale_factor=self.scale_factor)
+        self.postconv = nn.Conv1d(self.args.num_channel_wn, 1, kernel_size=1)
+        self.generator = WNBlocks(
+            in_channels=1,
+            hidden_channels=self.args.num_channel_wn,
+            kernel_size=self.args.kernel_size_wn,
+            dilation_rate=self.args.dilation_rate_wn,
+            num_blocks=self.args.num_blocks_wn,
+            num_layers=self.args.num_layers_wn,
+        )
+        self.waveform_disc = MelganMultiscaleDiscriminator(
+            downsample_factors=(2, 2, 2),
+            base_channels=16,
+            max_channels=1024,
+        )
+        self.spectral_disc = SpectralDiscriminator()
+
+    def init_from_config(config: Coqpit):
+        from TTS.utils.audio import AudioProcessor
+        ap = AudioProcessor.init_from_config(config)
+        return BWE(config, ap)
+
+    def gen_forward(self, x):
+        x = self.upsample(x)
+        x = self.generator(x)
+        x = self.postconv(x)
+        return {
+            "y_hat": x,
+        }
+
+    def disc_forward(self, x):
+        scores, feats = self.waveform_disc(x)
+        score, feats_ = self.spectral_disc(x)
+        scores.append(score)
+        return scores, feats + feats_
+
+    def forward(self, x):
+        return self.gen_forward(x)
+
+    def inference(self, x):
+        x = x.unsqueeze(1)
+        x = self.gen_forward(x)
+        return x
+
+    def train_step(self, batch: dict, criterion: nn.Module, optimizer_idx: int):
+        outputs = {}
+        loss_dict = {}
+        x = batch["input_wav"].unsqueeze(1)
+        y = batch["target_wav"].unsqueeze(1)
+        lens = batch["target_lens"]
+
+        if optimizer_idx not in [0, 1]:
+            raise ValueError(" [!] Unexpected `optimizer_idx`.")
+
+        if optimizer_idx == 0:
+            y_hat = self.gen_forward(x)["y_hat"]
+            self.y_hat_g = y_hat
+            scores_fake, feats_fake, feats_real = None, None, None
+            if self.train_disc:
+                scores_fake, feats_fake = self.disc_forward(y_hat)
```

Comment:
I think that you need to detach the y_hat here.

Reply:
my bad, the  optimizer_idx ==  0 is the generator optimizer, so it is right. I thought that it was the discriminator optimizer.